src/asynchro.rs

```diff
@@ -139,556 +139,563 @@ pub struct SincFixedOut<T> {
-                if is_x86_feature_detected!("sse3") {
-                    return Box::new(SseInterpolator::<$ft>::new(
-                        sinc_len,
-                        oversampling_factor,
-                        f_cutoff,
-                        window,
-                    ));
-                }
-                #[cfg(all(target_arch = "aarch64", feature = "neon"))]
-                if is_aarch64_feature_detected!("neon") {
-                    return Box::new(NeonInterpolator::<$ft>::new(
-                        sinc_len,
-                        oversampling_factor,
-                        f_cutoff,
-                        window,
-                    ));
-                }
-                Box::new(ScalarInterpolator::<$ft>::new(
-                    sinc_len,
-                    oversampling_factor,
-                    f_cutoff,
-                    window,
-                ))
-            }
+pub fn make_interpolator<T>(
+    sinc_len: usize,
+    resample_ratio: f64,
+    f_cutoff: f32,
+    oversampling_factor: usize,
+    window: WindowFunction,
+) -> Box<dyn SincInterpolator<T>>
+where
+    T: Sample,
+{
+    let sinc_len = 8 * (((sinc_len as f32) / 8.0).ceil() as usize);
+    let f_cutoff = if resample_ratio >= 1.0 {
+        f_cutoff
+    } else {
+        f_cutoff * resample_ratio as f32
+    };
 
-            /// Perform cubic polynomial interpolation to get value at x.
-            /// Input points are assumed to be at x = -1, 0, 1, 2
-            fn interp_cubic(&self, x: $ft, yvals: &[$ft; 4]) -> $ft {
-                let a0 = yvals[1];
-                let a1 =
-                    -(1.0 / 3.0) * yvals[0] - 0.5 * yvals[1] + yvals[2] - (1.0 / 6.0) * yvals[3];
-                let a2 = 0.5 * (yvals[0] + yvals[2]) - yvals[1];
-                let a3 = 0.5 * (yvals[1] - yvals[2]) + (1.0 / 6.0) * (yvals[3] - yvals[0]);
-                let x2 = x * x;
-                let x3 = x2 * x;
-                a0 + a1 * x + a2 * x2 + a3 * x3
-            }
+    #[cfg(all(target_arch = "x86_64", feature = "avx"))]
+    if let Ok(interpolator) =
+        AvxInterpolator::<T>::new(sinc_len, oversampling_factor, f_cutoff, window)
+    {
+        return Box::new(interpolator);
+    }
 
-            /// Linear interpolation between two points at x=0 and x=1
-            fn interp_lin(&self, x: $ft, yvals: &[$ft; 2]) -> $ft {
-                (1.0 - x) * yvals[0] + x * yvals[1]
-            }
-        }
-    };
+    #[cfg(target_arch = "x86_64")]
+    if let Ok(interpolator) =
+        SseInterpolator::<T>::new(sinc_len, oversampling_factor, f_cutoff, window)
+    {
+        return Box::new(interpolator);
+    }
+
+    #[cfg(all(target_arch = "aarch64", feature = "neon"))]
+    if let Ok(interpolator) =
+        NeonInterpolator::<T>::new(sinc_len, oversampling_factor, f_cutoff, window)
+    {
+        return Box::new(interpolator);
+    }
+
+    Box::new(ScalarInterpolator::<T>::new(
```

Comment:
The benches are failing now because the simd interpolator constructors all return Result<Self, ..>, while the scalar one directly returns Self. The easiest fix is to let also that one return a Result, and then add an unwrap() in the benches.